### PR TITLE
Fix  #1090

### LINKED ISSUE #1090: Cleaned library references are not removed from the deps.json file
Currently, weavers that set `ShouldCleanReference` to `true` get their library assembly removed from the dll references, but those will still be listed in the deps.json file. Most of the time this isn't a problem, but issues can arise in some cases.

See https://github.com/Fody/Anotar/issues/656

The rest of the issue description consists mostly of notes for myself, detailing my findings so far. I'll submit a PR for this when I'll have more time to work on it.

---

In order to fix this, Fody needs to remove the library path from the `RuntimeCopyLocalItems` items. This will remove the `"runtime"` part of the package dependency.

For instance, it will change this:

```JSON
"Anotar.Serilog.Fody/6.0.0": {
  "dependencies": {
    "Fody": "6.6.0",
    "Serilog": "2.10.0"
  },
  "runtime": {
    "lib/netstandard2.1/Anotar.Serilog.dll": {
      "assemblyVersion": "6.0.0.0",
      "fileVersion": "6.0.0.0"
    }
  }
},
```

to this:

```JSON
"Anotar.Serilog.Fody/6.0.0": {
  "dependencies": {
    "Fody": "6.6.0",
    "Serilog": "2.10.0"
  }
},
```

This fixes the reported issue in Anotar.

Fody and its weavers will remain as *package* dependencies in the deps.json file, but that shouldn't be a problem as long as no runtime dependency is defined.

After removing all references to `Anotar.Serilog.dll` from inputs of the `GenerateDepsFile` task (by removing them from the following items: `ReferencePath`, `ResolvedCompileFileDefinitions`, `NativeCopyLocalItems`, `ResourceCopyLocalItems`), references to the `Anotar.Serilog.Fody` and `Fody` packages still remain in the deps.json file. I believe these come from the project.assets.json file, which we cannot modify. That should be harmless though.

Even more so, `Anotar.Serilog.Fody` has a dependency on `Serilog` and brings the reference to the consuming project, so this relationship should remain in the deps.json file.

### PR DESCRIPTION
**Pull request by tom-englert**
Cleaned library references are not removed from the deps.json file

### CODE REVIEW

**Comment by ltrzesniewski** [comment]
Oh, thanks! I'll try to review this tomorrow.

**Review by SimonCropp — APPROVED** [review]

File: FodyHelpers/BaseModuleWeaver.cs [hunk 1]
```diff
@@ -214,6 +214,15 @@ public virtual void WriteError(string message, MethodDefinition method)
...
         public List<string> ReferenceCopyLocalPaths { get; set; } = new List<string>();
 
+        /// <summary>
+        /// A list of all the runtime references marked as copy-local.
+        /// A copy of the contents of the @(RuntimeCopyLocalItems).
+        /// </summary>
+        /// <remarks>
+        /// This list will be actively synced back to the build system, i.e. adding or removing items from this list will modify the @(RuntimeCopyLocalItems) list of the current build.
+        /// </remarks>
+        public List<string> RuntimeCopyLocalPaths { get; set; } = new List<string>();
```
[the remark this answers is not in the document]
Comment (tom-englert) [reply FodyHelpers/BaseModuleWeaver.cs:224]: Weavers that have "ShouldCleanReference" set to false must be able to remove items, or like e.g. ILMerge.Fody weavers must be able to remove also other dependencies that are no longer needed.

I agree adding items is not a common use case here, but I didn't want to duplicate all the code just to be able to block this.

File: Fody/Fody.targets [hunk 2]
```diff
@@ -96,6 +99,31 @@
 
   </Target>
 
+  <Target
+      Name="FodyUpdateRuntimeCopyLocalItemsTarget"
+      AfterTargets="FodyUpdateReferenceCopyLocalFilesTarget"
```
[the remark this answers is not in the document]
Comment (tom-englert) [reply Fody/Fody.targets:104]: Aggree, I was thinking too linear here

File: FodyHelpers/BaseModuleWeaver.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ltrzesniewski) [reply FodyHelpers/BaseModuleWeaver.cs:224]: Ok, ILMerge.Fody is a good example. :+1:

**Pull request merged**